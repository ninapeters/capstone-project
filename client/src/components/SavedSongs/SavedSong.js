import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { useState } from 'react'
import AudioButton from '../Buttons/AudioButton'
import Button from '../Buttons/Button'
import { CrossIcon, ResetIcon, BinIcon } from '../Icons'

SavedSong.propTypes = {
  stopPlayingSongById: PropTypes.func,
  artist: PropTypes.string.isRequired,
  songTitle: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  deleteSavedSong: PropTypes.func,
  toggleCurrentSongId: PropTypes.func,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
}

export default function SavedSong({
  stopPlayingSongById,
  artist,
  songTitle,
  id,
  deleteSavedSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
}) {
  const [toBeDeleted, setToBeDeleted] = useState(false)

  return (
    <>
      {toBeDeleted || (
        <ListItem>
          <ButtonWrapper>
            <ButtonPrepareToDelete
              isDeleteButton
              id={id}
              onClick={() => prepareToDelete(id)}
              data-testid="prepare-delete-button"
            >
              <CrossIcon />
            </ButtonPrepareToDelete>
          </ButtonWrapper>
          <Content>
            <Artist>{artist}</Artist>
            <Song>{songTitle}</Song>
            <AudioButtonWrapper>
              <AudioButton
                isSmall
                onClick={() => toggleCurrentSongId(id)}
                isSongPlaying={isSongPlaying && currentSongId === id}
                currentSongId={currentSongId}
              />
            </AudioButtonWrapper>
          </Content>
        </ListItem>
      )}
      {toBeDeleted && (
        <Container>
          <ConfirmationWrapper>
            <p>Do you want to delete {songTitle}?</p>
            <ConfirmationButtonWrapper>
              <DeleteButton
                id={id}
                onClick={() => deleteSavedSong(id)}
                data-testid="delete-button"
              >
                <BinIcon />
              </DeleteButton>
              <ResetButton
                id={id}
                onClick={() => setToBeDeleted(false)}
                data-testid="reset-button"
              >
                <ResetIcon />
              </ResetButton>
            </ConfirmationButtonWrapper>
          </ConfirmationWrapper>
        </Container>
      )}
    </>
  )

  function prepareToDelete(id) {
    stopPlayingSongById(id)
    setToBeDeleted(true)
  }
}

const ListItem = styled.li`
  display: grid;
  column-gap: 8px;
  grid-template-columns: 48px auto;
  grid-template-rows: auto;
  padding: 0 22px;
`
const ButtonWrapper = styled.div`
  grid-row: 1/3;
  grid-column-start: 1;
`
const ButtonPrepareToDelete = styled(Button)`
  border-radius: 34px 10px 10px 34px;
  height: 100%;
  width: 48px;
`
const Content = styled.section`
  box-shadow: var(--shadow-light);
  border-radius: 10px 34px 34px 10px;
  display: grid;
  grid-template-columns: 80% auto;
  grid-template-rows: auto;
  padding: 2px 4px 2px 12px;
`
const Artist = styled.span`
  align-self: end;
  color: var(--primary-regular);
  font-size: 0.9em;
  font-weight: 300;
  text-transform: uppercase;
`
const Song = styled.span`
  align-self: start;
  color: var(--primary-regular);
  grid-row-start: 2;
  font-size: 0.9em;
  font-weight: 700;
  text-transform: uppercase;
  line-height: 1;
`
const AudioButtonWrapper = styled.div`
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
  align-self: center;
`
const Container = styled.div`
  padding: 0 22px;
  display: grid;
`
const ConfirmationWrapper = styled.div`
  p {
    align-self: start;
    color: var(--primary-regular);
    font-size: 0.9em;
    font-weight: 700;
    text-align: center;
  }
`
const ConfirmationButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
`
const DeleteButton = styled(Button)`
  height: 48px;
  width: 48px;
  fill: var(--cta-red);
`
const ResetButton = styled(Button)`
  height: 48px;
  width: 48px;
  fill: var(--spotify-green);
`
