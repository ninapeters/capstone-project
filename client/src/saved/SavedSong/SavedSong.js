import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { useState } from 'react'
import AudioButton from '../../app/AudioButton'
import Button from '../../app/Button'
import { CrossIcon, ResetIcon, BinIcon } from '../../app/Icons/Icons'

SavedSong.propTypes = {
  stopPlayingSongById: PropTypes.func,
  artist: PropTypes.string.isRequired,
  songTitle: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  deleteSavedSong: PropTypes.func,
  toggleCurrentSongId: PropTypes.func,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
}

export default function SavedSong({
  stopPlayingSongById,
  artist,
  songTitle,
  id,
  deleteSavedSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
}) {
  const [toBeDeleted, setToBeDeleted] = useState(false)

  return (
    <>
      {toBeDeleted || (
        <ListItem>
          <PrepareToDeleteContainer>
            <Button
              isNoCircle
              isSmall
              isRedIcon
              id={id}
              onClick={() => prepareToDelete(id)}
              data-testid="prepare-delete-button"
            >
              <CrossIcon />
            </Button>
          </PrepareToDeleteContainer>
          <Content>
            <Artist>{artist}</Artist>
            <Song>{songTitle}</Song>
            <AudioButtonWrapper>
              <AudioButton
                isSmall
                onClick={() => toggleCurrentSongId(id)}
                isSongPlaying={isSongPlaying && currentSongId === id}
                currentSongId={currentSongId}
              />
            </AudioButtonWrapper>
          </Content>
        </ListItem>
      )}
      {toBeDeleted && (
        <Container>
          <ConfirmationWrapper>
            <p>Do you want to delete {songTitle}?</p>
            <ConfirmationButtonWrapper>
              <Button
                isSmall
                isRedIcon
                id={id}
                onClick={() => deleteSavedSong(id)}
                data-testid="delete-button"
              >
                <BinIcon />
              </Button>
              <Button
                isSmall
                id={id}
                onClick={() => setToBeDeleted(false)}
                data-testid="reset-button"
              >
                <ResetIcon />
              </Button>
            </ConfirmationButtonWrapper>
          </ConfirmationWrapper>
        </Container>
      )}
    </>
  )

  function prepareToDelete(id) {
    stopPlayingSongById(id)
    setToBeDeleted(true)
  }
}

const ListItem = styled.li`
  display: grid;
  column-gap: 8px;
  grid-template-columns: 48px auto;
  grid-template-rows: auto;
  padding: 0 22px;
`
const PrepareToDeleteContainer = styled.section`
  box-shadow: var(--shadow-light);
  border-radius: 34px 10px 10px 34px;
  display: grid;
  grid-template-rows: auto;
  padding: 4px 2px;
  place-items: center;
`
const Content = styled.section`
  box-shadow: var(--shadow-light);
  border-radius: 10px 34px 34px 10px;
  display: grid;
  grid-template-columns: 80% auto;
  grid-template-rows: auto;
  padding: 4px 4px 4px 12px;
`
const Artist = styled.span`
  align-self: end;
  color: var(--primary-regular);
  font-size: 0.9em;
  font-weight: 300;
  text-transform: uppercase;
`
const Song = styled.span`
  align-self: start;
  color: var(--primary-regular);
  grid-row-start: 2;
  font-size: 0.9em;
  font-weight: 700;
  text-transform: uppercase;
  line-height: 1;
`
const AudioButtonWrapper = styled.div`
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
  align-self: center;
`
const Container = styled.div`
  padding: 0 22px;
  display: grid;
`
const ConfirmationWrapper = styled.div`
  p {
    align-self: start;
    color: var(--primary-regular);
    font-size: 0.9em;
    font-weight: 700;
    text-align: center;
  }
`
const ConfirmationButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
`
