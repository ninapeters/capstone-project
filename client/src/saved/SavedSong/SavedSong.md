One saved song:

```jsx
<SavedSong
  stopPlayingSongById={() => {}}
  artist="Unprocessed"
  songTitle="Real"
  id="a"
  deleteSavedSong={() => {}}
  toggleCurrentSongId={() => {}}
  isSongPlaying={false}
  currentSongId="a"
/>
```
