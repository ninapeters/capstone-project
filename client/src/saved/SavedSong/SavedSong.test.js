import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import SavedSong from './SavedSong'

describe('SavedSong', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls toggleCurrentSongId with the correct id', () => {
    const toggleCurrentSongIdMock = jest.fn()
    const { getByTestId } = render(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => toggleCurrentSongIdMock('a')}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const button = getByTestId('audio-button')
    user.click(button)
    expect(toggleCurrentSongIdMock).toHaveBeenCalledWith('a')
  })

  it('has delete button', () => {
    const { getByTestId } = render(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const Button = getByTestId('prepare-delete-button')
    expect(Button).toBeInTheDocument()
  })

  it('renders delete field', () => {
    const { getByTestId } = render(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const Button = getByTestId('prepare-delete-button')
    user.click(Button)
    expect(getByTestId('reset-button')).toBeInTheDocument()
    expect(getByTestId('delete-button')).toBeInTheDocument()
  })

  it("stops playing the current song, if it's the one to be deleted", () => {
    const stopPlayingSongByIdMock = jest.fn()
    const { getByTestId } = render(
      <SavedSong
        stopPlayingSongById={() => stopPlayingSongByIdMock('a')}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const Button = getByTestId('prepare-delete-button')
    user.click(Button)
    expect(stopPlayingSongByIdMock).toHaveBeenCalledWith('a')
  })

  it('switches the appearance of play and pause button by changing the state isSongPlaying', () => {
    const { getByTitle, queryByTitle, rerender } = render(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    expect(getByTitle('play')).toBeInTheDocument()

    rerender(
      <SavedSong
        stopPlayingSongById={() => {}}
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={true}
        currentSongId="a"
      />
    )
    expect(queryByTitle('play')).not.toBeInTheDocument()
    expect(getByTitle('pause')).toBeInTheDocument()
  })
})
