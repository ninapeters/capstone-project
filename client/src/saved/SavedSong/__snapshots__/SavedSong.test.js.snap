// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedSong renders correctly 1`] = `
.c7 {
  background: transparent;
  border: none;
  border-radius: 50%;
  box-shadow: none;
  fill: var(--primary-regular);
  height: 52px;
  padding: 16px;
  width: 52px;
}

.c7:focus {
  outline: 0;
}

.c2 {
  background: transparent;
  border: none;
  border-radius: 34px 10px 10px 34px;
  box-shadow: none;
  fill: var(--cta-red);
  height: 50px;
  padding: 16px;
  width: 50px;
}

.c2:disabled {
  background: var(--spotify-green-transparent);
  fill: var(--white);
}

.c2:focus {
  outline: 0;
}

.c0 {
  display: grid;
  -webkit-column-gap: 8px;
  column-gap: 8px;
  grid-template-columns: 52px auto;
  grid-template-rows: auto;
  padding: 0 22px;
}

.c1 {
  box-shadow: var(--shadow-light);
  border-radius: 34px 10px 10px 34px;
  display: grid;
  grid-template-rows: auto;
  padding: 4px 2px;
  place-items: center;
}

.c3 {
  box-shadow: var(--shadow-light);
  border-radius: 10px 34px 34px 10px;
  display: grid;
  grid-template-columns: 80% auto;
  grid-template-rows: auto;
  padding: 4px 4px 4px 12px;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  color: var(--primary-regular);
  font-size: 0.9em;
  font-weight: 300;
  text-transform: uppercase;
}

.c5 {
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  color: var(--primary-regular);
  grid-row-start: 2;
  font-size: 0.9em;
  font-weight: 700;
  text-transform: uppercase;
  line-height: 1;
}

.c6 {
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<li
  class="c0"
>
  <section
    class="c1"
  >
    <button
      class="c2"
      data-testid="prepare-delete-button"
      style="transform: none;"
    >
      <svg>
        x.svg
      </svg>
    </button>
  </section>
  <section
    class="c3"
  >
    <span
      class="c4"
    >
      Unprocessed
    </span>
    <span
      class="c5"
    >
      Real
    </span>
    <div
      class="c6"
    >
      <button
        class="c7"
        data-testid="audio-button"
        style="transform: none;"
      >
        <svg
          title="play"
        >
          play.svg
        </svg>
      </button>
    </div>
  </section>
</li>
`;
